### PR TITLE
Create fsstore from filesystem

### LINKED ISSUE #993: Cannot create FSStore for unlistable http URLs
#### Minimal, reproducible code sample, a copy-pastable example if possible

```python
import zarr
import fsspec

# public
url = 'https://g-c70289.0da32.08cc.data.globus.org/pyqg_ensemble/001.zarr'

# this errors
store = zarr.storage.FSStore(url)

# because of this
fs = fsspec.implementations.http.HTTPFileSystem()
fs.isdir(url) # -> False

# however, there is consolidated metadata
data = fs.get_mapper(url)['.zmetadata']
```

#### Problem description

We often store Zarr on HTTP servers that don't support directory listing. This works because Zarr consolidated metadata tells us everything we need to know about the data available, so no listing is required.

However, Zarr won't let me create such a Store, as shown in the example above.

There needs to be a workaround to allow unlistable http stores with consolidated metadata. 

#### Version and installation information

Please provide the following:

* zarr 2.11.0
* fsspec 2022.02.0

This came up in https://github.com/pangeo-forge/pangeo-forge-recipes/issues/222

### PR DESCRIPTION
**Pull request by rabernat**
This PR adds a new argument to `FSStore` which allows one to pass an existing fsspec filesystem object from which to generate the FSStore. The need to initialize an FSStore from an existing filesystem came up in https://github.com/pangeo-forge/pangeo-forge-recipes/issues/254. It seems like this is something people may want to do more generally.

Also fixes #993 by simply removing the check for list-ability when initializing an FSStore.

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [x] GitHub Actions have all passed
* [x] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: zarr/storage.py
Comment (martindurant) [inline zarr/storage.py:1094]: Perhaps an even simpler alternative is if the input is already an FSMap? We could handle both those cases. 

Is the intent that this will be called from zarr.open* or xarray, or that calling code will instantiate this themselves before calling zarr.open(fsstore) ?

File: zarr/storage.py
Comment (martindurant) [inline zarr/storage.py:1110]: There are a couple of extra options that FSMap takes over what a FileSystem takes (check, create, exceptions) that could go here.

File: zarr/tests/test_storage.py
Comment (martindurant) [inline zarr/tests/test_storage.py:1138]: I suppose we should test these in combination with top level zarr objects.

File: zarr/storage.py
Comment (rabernat) [reply zarr/storage.py:1094]: Ok so it seems like there are three basic cases we need to cover:
1. The user wants to specify a URL (including protocol)
2. The user already has a filesystem initialized and wants to use that to create the FSStore
3. The user already has an FSMap object and want to use that to create the FSStore

I am already uncomfortable with the overloading of the constructor on this class to handle two different scenarios. If we expand it to cover 3, it will get even more messy. A better pattern would be

```python
class FSStore(Store):

    @classmethod
    def from_url(cls, url):
        # figure out fs and path
        return cls(path, fs)

    @classmethod
    def from_fsmap(cls, fsmap):
        # figure out fs and path
        return cls(path, fs)

    def __init__(self, path, fs):
        ...
```
However, this would change the API and break lots of code.

More generally, looking at this code has made me realize that the interface between zarr.storage and FSSpec is not very clean. `FSStore` reimplements and extends many of the MutableMapping methods that `FSMap` implements. `FSStore` calls both `self.map` methods and `self.fs` methods (some of them private 😬 ).

File: zarr/storage.py
Comment (rabernat) [reply zarr/storage.py:1110]: Are these the only possible options - https://github.com/fsspec/filesystem_spec/blob/201555581a9f0c0de810efc372704db952397330/fsspec/mapping.py#L35 - `check=False, create=False, missing_exceptions=None`?

File: zarr/tests/test_storage.py
Comment (rabernat) [reply zarr/tests/test_storage.py:1138]: Where do those tests belong?

File: zarr/storage.py
Comment (martindurant) [reply zarr/storage.py:1094]: I agree with all of that. As you have seen, this is a consequence of the evolution of the storage class in zarr itself - originally it was to be a MutableMap alone, so that could go into fsspec and be generally useful, but then stores started to gain additional capabilities. And here we are.

I am *not* uncomfortable with the class methods. I think we can have `__init__` stay the same (and not break code), but add `from_filesystem(fs, path, args)`, `from_mapper(mapper)` class methods. Either the calling code will determine which to call, or we add another helper to decide?

File: zarr/tests/test_storage.py
Comment (martindurant) [reply zarr/tests/test_storage.py:1138]: I would say right here. I haven't looked to see what other tests do for full integration/roundtrips, but fsspec is certainly a bit special.

**Comment by rabernat** [comment]
I would like to revive and move forward this PR. Martin, you gave a few helpful comments and ideas above. Would you mind turning these into a more formal review in terms of what you think is _needed_ vs _interesting possibility_.

What we have here is a tightly scoped and tested PR to unblock a rather serious upstream problem. If there is no major technical debt incurred by this PR, could we consider merging as is (modulo doc updates from checklist) rather than expanding scope? If not, please let us know what else would be required.

Also pinging @grlee77 for potential interaction with ongoing refactor of storage classes (#600).

File: zarr/storage.py
Comment (grlee77) [reply zarr/storage.py:1094]: Yeah, if `__init__` stayed the same then the rest of this seems pretty non-disruptive to me. I do think a classmethod approach looks nice.

File: zarr/tests/test_storage.py
Comment (grlee77) [reply zarr/tests/test_storage.py:1138]: There are tests with higher-level zarr `Array` objects in `test_core.py` (e.g. `TestArrayWithFSStore`) and for `Group` in `test_hierarchy.py` (see `TestGroupWithFSStore`). Currently those just create an FSStore using a path string as returned by `tempfile.mkdtemp()`

**Comment by grlee77** [comment]
> Also pinging @grlee77 and @joshmoore for potential interaction with ongoing refactor of storage classes (#789).

Thanks @rabernat, I commented above. If the `__init__` signature does not change then this should be pretty easy to merge into the work there.

File: zarr/storage.py
Comment (rabernat) [reply zarr/storage.py:1094]: Thanks a lot for the feedback Greg! Just to be totally explicit, is adding an additional argument  to `__init__` ( as I have have done here with `fs`), while keeping the other arguments backwards compatible, considered a "disruptive change"? Or is this PR already sufficiently non-disruptive.

Comment (grlee77) [reply zarr/storage.py:1094]: Yeah, what is currently in the PR seems fine to me. It should be easy to just merge this into the v3 branch with minimal conflicts or extra work.

**Comment by joshmoore** [comment]
Thanks, @grlee77. From my side, that was the only concern in terms of impact on existing work, @rabernat. @martindurant's point about the general API "burden" seems to be outstanding. Is it that a concrete alternative that doesn't break the API needs proposing from your side, @rabernat?

A side note that testing is currently blocked due to GHA weirdness. @jakirkham and I will try to get that sorted ASAP.

File: zarr/storage.py
Comment (rabernat) [inline zarr/storage.py:1332]: To fix #993 I have simply removed this check...

File: zarr/tests/test_storage.py
Comment (rabernat) [inline zarr/tests/test_storage.py:1253]: ...and its corresponding test.

This test did not do anything other than test the error above. `self.create_store` does not actually attempt to write any data or list the store.

With the check gone, users will now see an error from fsspec if they try to do an operation that requires listing on an unlistable FSSpec store, as they would already on a generic MutableMapping-based store.

File: zarr/storage.py
Comment (rabernat) [reply zarr/storage.py:1310]: As the current [test failure](https://github.com/zarr-developers/zarr-python/runs/5820026384?check_suite_focus=true#step:7:1634) shows, it is not supported by `LocalFileSystem`.

File: zarr/storage.py
Comment (rabernat) [reply zarr/storage.py:1285]: Now that https://github.com/fsspec/filesystem_spec/pull/939 has gone in, it should be possible to create an FSStore from an FSMap. But let's address that in a future PR, since that functionality is still unreleased.

File: docs/release.rst
Comment (rabernat) [inline docs/release.rst:28]: Although 2.11.3 has been released, I noted that the docs had not been updated, so I added it here.

It seems we have skipped release 2.11.2?

Comment (joshmoore) [reply docs/release.rst:28]: Those releases are coming from the `2_11` branch. (See the related https://github.com/zarr-developers/zarr-python/pull/898#issuecomment-1076912502)

I assume you intend this for release as a 2.11.x?

Comment (martindurant) [reply docs/release.rst:28]: Please ping this thread with any decision on which branch to merge into

Comment (joshmoore) [reply docs/release.rst:28]: @martindurant : @rabernat seemed to think it was ok to hold off on @grlee77's upcoming v3 configuration fix, i.e. this can stay on the branch it is and be released as something `>2.11.x`. If it looks like we want it in a quick `~=2.11.x` release, I can attempt the backport.

Comment (rabernat) [reply docs/release.rst:28]: I will do whatever the other devs recommend here (either backport to 2.11.x or put this in 2.12.0)

I am confused about how we will maintain a consistent changelog for these two branches. So I am standing by for advice.

**Comment by rabernat** [comment]
I feel confused about how to move forward with this PR in terms of how to update the changelog (see https://github.com/zarr-developers/zarr-python/pull/911#discussion_r848818092). Please help me 🙏. I would love to see this merged soon as I believe it is finished. It is not starting to accumulate conflicts.

**Comment by joshmoore** [comment]
Sorry for having left you hanging. I think either will be fine @rabernat. Either just add your item to unreleased or copy over from the 2 11 branch. Either way feel free to get this in and I will double check before the next release. (Of whatever branch)

**Review by joshmoore — APPROVED** [review]
I've hopefully handled the conflicts with Greg's v3 work. Once this goes green, I'll merge and update the release notes to include previous versions.

**Comment by joshmoore** [comment]
It looks like the semantics of the `update()` method have changed with the refactoring for v3:

```
pytest zarr/tests/test_convenience.py -svx -k2

...snip...

>           store_to_open.update(store)  # copy original store to new unlistable store

zarr/tests/test_convenience.py:305:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <zarr.storage.FSStore object at 0x1ba010970>, other = '/var/folders/z5/txc_jj6x5l5cm81r56ck1n9c0000gn/T/tmp0gy4ji7i', kwds = {}

    def update(self, other=(), /, **kwds):
        ''' D.update([E, ]**F) -> None.  Update D from mapping/iterable E and F.
            If E present and has a .keys() method, does:     for k in E: D[k] = E[k]
            If E present and lacks .keys() method, does:     for (k, v) in E: D[k] = v
            In either case, this is followed by: for k, v in F.items(): D[k] = v
        '''
        if isinstance(other, Mapping):
            for key in other:
                self[key] = other[key]
        elif hasattr(other, "keys"):
            for key in other.keys():
                self[key] = other[key]
        else:
>           for key, value in other:
E           ValueError: not enough values to unpack (expected 2, got 1)
```


That doesn't appear to be related to the type hierarchy:
```
2.11.2: FSStore.__mro__: (<class 'zarr.storage.FSStore'>, <class 'zarr._storage.store.Store'>, <class 'zarr._storage.store.BaseStore'>, <class 'collections.abc.MutableMapping'>, <class 'collections.abc.Mapping'>, <class 'collections.abc.Collection'>, <class 'collections.abc.Sized'>, <class 'collections.abc.Iterable'>, <class 'collections.abc.Container'>, <class 'object'>)
2.12.x: FSStore.__mro__: (<class 'zarr.storage.FSStore'>, <class 'zarr._storage.store.Store'>, <class 'zarr._storage.store.BaseStore'>, <class 'collections.abc.MutableMapping'>, <class 'collections.abc.Mapping'>, <class 'collections.abc.Collection'>, <class 'collections.abc.Sized'>, <class 'collections.abc.Iterable'>, <class 'collections.abc.Container'>, <class 'object'>)
```
Still digging in case anyone has an idea (cc: @grlee77)

**Comment by rabernat** [comment]
Thanks for finishing this up Josh! I really appreciate it!

**Pull request merged**
